go/marshal/encode.go

```diff
@@ -223,18 +260,29 @@ func structEncoder(t reflect.Type, parentStructTypes []reflect.Type) encoderFunc
 
 	parentStructTypes = append(parentStructTypes, t)
 	fields, structType, originalFieldIndex := typeFields(t, parentStructTypes)
-	if structType != nil {
+
+	// It won't be possible to use the pre-computed struct type if it implements
+	// MarshalOptionsGetter, because the struct name can change at runtime.
+	isOptsGetter := t.Implements(optionsGetterType)
+
+	if originalFieldIndex != nil {
+		// Slowest path - we are extending some other struct. We need to start with the
```

Comment:
I honestly don't really understand this code, or whether this "original" thing is affected by the marshal options. It seems to me like it shouldn't (?).